app/src/main/java/ch/epfl/cs311/wanderwave/viewmodel/PlayerViewModel.kt

```diff
@@ -0,0 +1,86 @@
+package ch.epfl.cs311.wanderwave.viewmodel
+
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import ch.epfl.cs311.wanderwave.model.data.Track
+import ch.epfl.cs311.wanderwave.model.spotify.SpotifyController
+import dagger.hilt.android.lifecycle.HiltViewModel
+import javax.inject.Inject
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.SharingStarted
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.combine
+import kotlinx.coroutines.flow.stateIn
+import kotlinx.coroutines.launch
+
+@HiltViewModel
+class PlayerViewModel @Inject constructor(val spotifyController: SpotifyController) : ViewModel() {
+  private val _playerState = spotifyController.playerState()
+  private val _expandedState = MutableStateFlow(false)
+
+  private var _uiState =
+      combine(_playerState, _expandedState) { playerState, expandedState ->
+        if (playerState == null) {
+          UiState(expanded = expandedState)
+        } else
+            UiState(
+                /* TODO: I _think_ that the track uri includes "spotify:track:" at the beginning, so
+                 *   need to remove that. */
+                track = playerState.track?.let { Track(it.uri, it.name, it.artist.name) },
+                isPlaying = !playerState.isPaused,
+                repeatMode = playerState.playbackOptions.repeatMode != LoopMode.NONE.ordinal,
+                isShuffling = playerState.playbackOptions.isShuffling,
+                expanded = expandedState)
+      }
+  val uiState: StateFlow<UiState> =
+      _uiState.stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), UiState())
+
+  init {
+    spotifyController.setOnTrackEndCallback { skipForward() }
+  }
+
+  fun collapse() {
+    _expandedState.value = false
+  }
+
+  fun expand() {
+    _expandedState.value = true
+  }
+
+  fun resume() {
+    spotifyController.resumeTrack()
+  }
+
+  fun pause() {
+    spotifyController.pauseTrack()
+  }
+
+  fun skipForward() {
+    viewModelScope.launch { spotifyController.skip(1) }
+  }
+
+  fun skipBackward() {
+    viewModelScope.launch { spotifyController.skip(-1) }
+  }
+
+  fun toggleShuffle() {
```

Comment:
maybe remove the commented code?

Reply:
Well, rather than remove it, I probably need to implement it :)
Done, the Shuffling logic is now done in the SpotifyController.